### ISSUE TITLE
to_crs creates different result from pyproj.transform

### ISSUE THREAD

**Issue by Marchen** [geopandas/geopandas#1245]
I'm not sure this is an issue for GeoPandas or pyproj, but parhaps due to `always_xy` option for `pyproj.transform`, `to_crs` creates different result from `pyproj.transform`.

# Repro

## Python and package versions

```python
import sys

import pyproj
import geopandas

print(sys.version)
print(pyproj.__version__)
print(geopandas.__version__)
```

>3.8.0 | packaged by conda-forge | (default, Nov 22 2019, 19:11:19) 
>[Clang 9.0.0 (tags/RELEASE_900/final)]

>2.4.2.post1

>0.6.2

## Reproject by pyproj.transform

```python
x = -567637.0
y = -56204.0

# EPSG 4612: JGD2000, geographical, unit: degree
# https://epsg.io/4612
EPSG4612 = pyproj.Proj("EPSG:4612")
# RPSG 2443: JGD2000 / Japan Plane Rectangular CS I, projected, unit: meter
# https://epsg.io/2443
EPSG2443 = pyproj.Proj("EPSG:2443")

print(pyproj.transform(EPSG2443, EPSG4612, x, y))
```

>(27.878013118155934, 128.92918943580798)

Here, 27.8780 is latitude and 128.929 is longitude.

## Reproject by GeoDataFrame.to_crs

```python
gdf = geopandas.GeoDataFrame(
    geometry = geopandas.points_from_xy(x = [x], y = [y]),
    crs = "EPSG:2443"
)
print(gdf.to_crs("EPSG:4612"))
```

>geometry
>0  POINT (123.47406 32.34878)

## Swapping x and y produce correct result

```python
gdf = geopandas.GeoDataFrame(
    geometry = geopandas.points_from_xy(x = [y], y = [x]),
    crs = "EPSG:2443"
)
print(gdf.to_crs("EPSG:4612"))
```

>geometry
>0  POINT (128.92919 27.87801)

## Result produced by to_crs is similar for the one produced by transform with `always_xy = True`

```python
print(pyproj.transform(EPSG2443, EPSG4612, x, y, always_xy=True))
```

>(123.4740611549679, 32.34878175793676)

**Comment by martinfleis** [comment]
cc @snowman2

**Comment by snowman2** [comment]
It is indeed the always xy option: https://github.com/geopandas/geopandas/pull/1122

I am wondering if exposing the always_xy option in  the to_crs method would useful?

**Comment by Marchen** [comment]
I think exposing the always_xy option would be useful and better than something like points_from_xy(x  = y, y = x).

**Comment by martinfleis** [comment]
> I am wondering if exposing the always_xy option in the to_crs method would useful?

@snowman2 sure, that is a possibility, but how come the result of transformation is wrong in the first place? That is something which should not happen.

**Comment by Marchen** [comment]
@martinfleis this is because axis ordering of PROJ is not consistent and setting always_xy for transform swaps x and y axes.
https://proj.org/faq.html#why-is-the-axis-ordering-in-proj-not-consistent

**Comment by Marchen** [comment]
But if geopandas automatically handle axis order correctly, I think it is best for users.

**Comment by jorisvandenbossche** [comment]
I am not sure we should expose `always_xy` in `to_crs` in geopandas. A lot of other functionality in GeoPandas assumes that your coordinates in the geometry objects are in x/y order, and I don't think we should support any other order.

**Comment by martinfleis** [comment]
Agree with @jorisvandenbossche here. However, we should ideally raise a warning if the order of source/target CRS is y/x and properly document this.

**Comment by jorisvandenbossche** [comment]
Better documentation for sure, but in which cases would you like to see a warning? Eg EPSG:4326 is in y/x order, but I don't think we should warn each time when that epsg number is used?

**Comment by martinfleis** [comment]
In cases like the one above, which seems to produce incorrect results. I don't know what should be the rule here, as my knowledge of these CRS corner cases is fairly limited. But if I understand correctly, `to_crs` in the example above returns incorrect transformation. User should be at least wanted that something like that could have happened.

**Comment by jorisvandenbossche** [comment]
But then we first need to understand what is going wrong in the above example.

I didn't look into very much detail, but if the CRS is lat/lon or y/x, shouldn't those be passed as such to transform? Swapping x and y, you get the same result with pyproj as with geopandas:

```
In [5]: print(pyproj.transform(EPSG2443, EPSG4612, y, x)) 
(32.34878175793676, 123.4740611549679)
```

**Comment by martinfleis** [comment]
The original CRS is using y, x order. By setting `always_xy` to True, pyproj expects that both projections are using x, y. 

> If true, the transform method will accept as input and return as output
            coordinates using the traditional GIS order, that is longitude, latitude
            for geographic CRS and easting, northing for most projected CRS.

That essentially switches coordinates and hence returns incorrect coordinates. 

As geopandas always expects x, y, and does not support any other order, I would raise a warning during assignment of such a CRS, i.e. during `read_file` and then during `to_crs`, which will always return incorrect result.

I am a bit confused about EPSG:4326. It indeed uses lat/lon (y/x) according to specification, but `naturalearth_lowres` which is in 4326 is x/y...

**Comment by snowman2** [comment]
> I am a bit confused about EPSG:4326. It indeed uses lat/lon (y/x) according to specification, but naturalearth_lowres which is in 4326 is x/y...

That is because historically GIS programs have ignored axis order. That file and others likely suffer from the same issue of ignoring axis order. The always_xy option preserves the past behavior.

**Comment by martinfleis** [comment]
In that case I would expand our documentation to talk about these issues and expose `always_xy`. The purpose of `always_xy` in that case should be to allow for transformation from y/x CRS to x/y. Assuming your dataset is originally in y/x like the one in the example above. I would not support y/x across geopandas, but just give users a way how to reproject their data into supported CRS correctly.

**Comment by snowman2** [comment]
That sounds like a change to add after #1101 is merged in.

**Comment by gepcel** [comment]
I'm confused here. From point `longitude=-567637.0, latitude=-56204.0` in EPSG2443, transformed to EPSG4612, the following two results (result_A, and result_B):

**result_A** from pyproj.transform:
```
longitude = 128.929
latitude = 27.8780
```

**result_B** from gpd.to_crs:
```
longitude = 123.47406
latitude = 32.34878
```

The correct result is [`result_B`](https://epsg.io/transform#s_srs=2443&t_srs=4612&x=-567637.0000000&y=-56204.0000000), right?

**Comment by snowman2** [comment]
Depends on your definition of correct. I would recommend reading:
https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6

**Comment by snowman2** [comment]
Looks
It looks like the input projection has the northing coordinate first.
```python
>>> from pyproj import CRS
>>> CRS("EPSG:2443")
<Projected CRS: EPSG:2443>
Name: JGD2000 / Japan Plane Rectangular CS I
Axis Info [cartesian]:
- X[north]: Northing (metre)
- Y[east]: Easting (metre)
Area of Use:
- name: Japan - zone I
- bounds: (128.17, 26.96, 130.46, 34.74)
Coordinate Operation:
- name: Japan Plane Rectangular CS zone I
- method: Transverse Mercator
Datum: Japanese Geodetic Datum 2000
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich
```
And the output has the northing coordinate first.
```python
>>> CRS("EPSG:4612")
<Geographic 2D CRS: EPSG:4612>
Name: JGD2000
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: Japan
- bounds: (122.38, 17.09, 157.65, 46.05)
Datum: Japanese Geodetic Datum 2000
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich
```

So, you need to input/output with north first and expect lat first:
```python
>>> from pyproj import Transformer
>>> north = -56204.0
>>> east = -567637.0
>>> trans = Transformer.from_crs(2443, 4612)
>>> lat, lon = trans.transform(north, east)
>>> lat, lon
(32.34878175793677, 123.4740611549679)
```
However, `geopandas` uses `always_xy`, so it expects input x first and output x first always:
```python
>>> transxy = Transformer.from_crs(2443, 4612, always_xy=True)
>>> lon, lat = transxy.transform(east, north)
>>> lon, lat
(123.4740611549679, 32.34878175793677)
```

Based on this, `result_B` seems to be correct.

**Comment by gepcel** [comment]
Thanks, @snowman2. I guess my confusion comes from that I always assume x equals longitude equals east, and y equals latitude equals north, despite whichever axis order in using.

By "correct", I mean after transformed the two points should be the same location on a map, or on the earth, right?

**Comment by snowman2** [comment]
> By "correct", I mean after transformed the two points should be the same location on a map, or on the earth, right?

Yeah, you're right. I originally skimmed your question :upside_down_face: - bad habit of mine. I thought the order of the output was swapped as is shown in the example I provided. But, looking at it again, the values do differ. So only one of the two could have been correct.
